app/controllers/lti_controller.rb

```diff
@@ -25,31 +26,125 @@ def get_canvas_config
               {
                 text: I18n.t('lti.launch'),
                 placement: 'course_navigation',
-                target_link_uri: root_url,
+                target_link_uri: lti_redirect_login_url,
                 canvas_icon_class: 'icon-lti',
                 default: 'disabled',
                 visibility: 'admins',
-                windowTarget: '_blank'
+                windowTarget: '_blank',
+                custom_fields: {
+                  user_id: '$Canvas.user.id',
+                  course_id: '$Canvas.course.id',
+                  course_name: '$Canvas.course.name'
+                }
               }
             ]
           }
         }
       ],
-      public_jwk_url: lti_public_jwk_url
+      public_jwk_url: lti_public_jwk_url,
+      custom_fields: {
+        user_id: '$Canvas.user.id',
+        course_id: '$Canvas.course.id',
+        course_name: '$Canvas.course.name'
+      }
     }
+
     render json: config.to_json
   end
 
   def launch
-    # TODO: implement this function
-    render json: { state: 'success' }
+    if params[:client_id].blank? || params[:login_hint].blank? ||
+      params[:target_link_uri].blank? || params[:lti_message_hint].blank?
+      head :unprocessable_entity
+      return
+    end
+    nonce = rand(10 ** 30).to_s.rjust(30, '0')
+    session[:client_id] = params[:client_id]
+    session[:nonce] = nonce
+    auth_params = {
+      scope: 'openid',
+      response_type: 'id_token',
+      client_id: params[:client_id],
+      redirect_uri: params[:target_link_uri],
+      lti_message_hint: params[:lti_message_hint],
+      login_hint: params[:login_hint],
+      response_mode: 'form_post',
+      nonce: nonce,
+      prompt: 'none',
+      state: session.id
+    }
+    referrer = URI(request.referer)
+
+    # TODO: generalize this to platforms other than canvas.
+    auth_request_uri = URI("#{referrer.scheme}://#{referrer.host}:#{referrer.port}/api/lti/authorize_redirect")
+
+    http = Net::HTTP.new(auth_request_uri.host, auth_request_uri.port)
+    req = Net::HTTP::Post.new(auth_request_uri)
+    req.set_form_data(auth_params)
+
+    res = http.request(req)
+    location = res['location']
+    redirect_to location
+  end
+
+  def redirect_login
+    referrer_uri = URI(request.referer)
+    known_jwks = %w[q.utoronto.ca canvas.instructure.com] # only supporting canvas right now
```

Comment:
1. Let's make this check a `before_action` filter for all methods in this controller.
2. Make this list a configuration setting (so that we can use `host.docker.internal` in development.